cetmix_tower_server/models/cx_tower_key_mixin.py

```diff
@@ -0,0 +1,59 @@
+from odoo import api, fields, models
+
+
+class CxTowerKeyMixin(models.AbstractModel):
+    _name = "cx.tower.key.mixin"
+    _description = "Mixin for managing secrets"
+
+    secret_ids = fields.Many2many(
+        comodel_name="cx.tower.key",
+        compute="_compute_secret_ids",
+        compute_sudo=True,
+        string="Secrets",
+    )
+
+    @api.depends("code")
+    def _compute_secret_ids(self):
+        """
+        Compute the secret IDs based on the references found in the code field.
+
+        This method updates the secret_ids Many2many field by extracting secret
+        references from the code field. If no code is present, the field is cleared.
+        It ensures updates are only triggered when there are differences between
+        the current and new secret IDs.
+        """
+        for record in self:
+            if record.code:
+                # Extract the new set of secret IDs
+                new_secret_ids = set(self._extract_secret_ids(record.code))
+
+                # Get the current secret IDs
+                current_secret_ids = set(record.secret_ids.ids)
+
+                # Update only if there's a difference
+                if new_secret_ids != current_secret_ids:
+                    record.secret_ids = [(6, 0, list(new_secret_ids))]
+            else:
+                record.secret_ids = [(5, 0, 0)]
+
+    @api.model
+    def _extract_secret_ids(self, code):
+        """
+        Extract secret IDs based on references found in the given `code`.
+
+        Args:
+            code: Text containing potential secret references.
+
+        Returns:
+            list: List of secret IDs corresponding to the references in `code`.
+        """
+        key_model = self.env["cx.tower.key"]
+        key_strings = key_model._extract_key_strings(code)
+
+        key_refs = []
+        for key_string in key_strings:
+            key_parts = key_model._extract_key_parts(key_string)
+            if key_parts:
+                key_refs.append(key_parts[1])
+
+        return key_model.search([("reference", "in", key_refs)]).ids
```

Comment:
We should need in mind, that `reference` field in the `cx.tower.key` model can [contain](https://github.com/cetmix/cetmix-tower/blob/e80208ca0882530797e1471fe7a6095097dcb5a4/cetmix_tower_server/models/cx_tower_key.py#L49) non unique values.
So we should check only those keys, that don't belong to any partner or server.

Reply:
thanks, I'll check and fix it